File: python-solutions/meeting-rooms-2.py
from typing import List


class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        start_times = sorted(i[0] for i in intervals)
        end_times = sorted(i[1] for i in intervals)
        used_rooms = 0

        end_pointer = 0

        for i in range(len(intervals)):
            if (start_times[i] >= end_times[end_pointer]):
                used_rooms -= 1
                end_pointer += 1

            used_rooms += 1

        return used_rooms

File: javascript-solutions/meeting-rooms-2.js
/**
 * @param {number[][]} intervals
 * @return {number}
 */
const minMeetingRooms = function (intervals) {
	if (!intervals.length) return 0;
	const startTimes = intervals
		.sort((a, b) => a[0] - b[0])
		.map((interval) => interval[0]);
	const endTimes = intervals
		.sort((a, b) => a[1] - b[1])
		.map((interval) => interval[1]);
	let usedRooms = 0;

	let endPointer = 0;

	intervals.forEach((interval, i) => {
		// The start time of this meeting starts later than or at the ending time of the least recently started meeting
		// which means a room is now free to be reused
		if (startTimes[i] >= endTimes[endPointer]) {
			// Free up a room
			usedRooms -= 1;
			// Move to the next least recently used meeting end time
			endPointer += 1;
		}
		// Increase the number of rooms, this will be cancelled out when the previous if statement equates to true
		usedRooms += 1;
	});
	return usedRooms;
};
